gl.cpp

```diff
@@ -354,29 +370,36 @@ GLFix nglZBufferAt(const unsigned int x, const unsigned int y)
     return z_buffer[x + y * SCREEN_WIDTH];
 }
 
+constexpr GLFix ABS(GLFix a) {
+    return a >= GLFix(0) ? a : -a;
+}
+
 //Doesn't interpolate colors even if enabled
 void nglDrawLine3D(const VERTEX *v1, const VERTEX *v2)
 {    
-    //TODO: Z-Clipping
+    // Z-Clipping!
+    if(v1->z < near_plane || v2->z < near_plane)
+        return;
 
     VERTEX v1_p = *v1, v2_p = *v2;
     nglPerspective(&v1_p);
     nglPerspective(&v2_p);
 
     const GLFix diff_x = v2_p.x - v1_p.x;
-    const GLFix dy = (v2_p.y - v1_p.y) / diff_x;
+    const GLFix diff_y = v2_p.y - v1_p.y;
 
     const COLOR c = v1_p.c;
 
     //Height > width? -> Interpolate X
-    if(dy > GLFix(1) || dy < GLFix(-1))
+    //if(dy > GLFix(1) || dy < GLFix(-1))
+
+    // if check passes diff_y should always be non zero
```

Comment:
You can be more confident here:

```suggestion
    // if check passes, diff_y is non zero
```